lib/cache/cache.go

```diff
@@ -16,6 +20,7 @@ type Cache struct {
 	expireRunning  bool
 	expireDuration time.Duration // expire the cache entry when it is older than this
 	expireInterval time.Duration // interval to run the cache expire
+	err            error         // any error, e.g. while dropping backends which can shutdown
```

Comment:
Instead of storing the error here, I'd suggest adding a value

```go
finalize func(value interface{})
```

and a method

```go
func (c *Cache) SetFinalizer(finalize func(value interface{}))
```

Which can be used to have a callback for finalizing objects.

That will look a lot like `tryShutdown` except it will use `fs.Errorf` to log each error, and it will call `fs.CountError` to set the global error flag.

func (c *Cache) AddFinalizer(func()